### DISCUSSION TITLE
Is it safe to assume that Walk will go through every file in a subdirectory first before moving on to another one?

### DISCUSSION THREAD

**Discussion by ItsSunnyMonster** [BurntSushi/ripgrep#2847] [Q&A] [answered]
Basically what the title says. In the `ignore` crate, the `Walk` struct is described as "a recursive directory iterator over file paths in one or more directories." Can I safely assume that it will return all files from a subdirectory first before returning items from another directory on all platforms?

**Comment by BurntSushi** [comment, marked as answer]
I don't think so, no. And you especially can't when using `WalkParallel`.

I believe single threaded will use depth first traversal. So you shouldn't even be seeing the behavior you're asking for (breadth first search) in practice...

**Reply by ItsSunnyMonster** [reply]
Thank you! I think I wasn't clear enough when I asked about this but looking at how depth-first search works it is exactly the behaviour I want.

I suppose the confusion came from what counts as files from a subdirectory. I needed it to recursively return all files from the subdirectory and any of *its* subdirectories first before it moves on to another subdirectory.

But thank you again for responding so quickly!
